File: license-service/src/kafka_producer.py
import json
import os
from datetime import datetime
from kafka import KafkaProducer
from kafka.errors import KafkaError
import logging

logger = logging.getLogger(__name__)


class KafkaProducerService:
    def __init__(self):
        kafka_broker = os.getenv('KAFKA_BOOTSTRAP_SERVERS', 'kafka:9093')
        self.topic = os.getenv('KAFKA_LICENSE_EVENTS_TOPIC', 'license-events')
        client_id = os.getenv('KAFKA_CLIENT_ID', 'license-service')
        self.event_source = os.getenv('KAFKA_EVENT_SOURCE', 'license-service-legacy')
        
        try:
            self.producer = KafkaProducer(
                bootstrap_servers=[kafka_broker],
                client_id=client_id,
                value_serializer=lambda v: json.dumps(v).encode('utf-8'),
                key_serializer=lambda k: k.encode('utf-8') if k else None
            )
            logger.info(f'Kafka producer initialized with broker: {kafka_broker}')
        except Exception as e:
            logger.error(f'Failed to initialize Kafka producer: {e}')
            self.producer = None

    def publish_event(self, event_type: str, user_id: str, payload: dict):
        """Publish a license event to Kafka"""
        try:
            event = {
                'eventType': event_type,
                'userId': user_id,
                'payload': payload,
                'timestamp': datetime.utcnow().isoformat() + 'Z',
                'source': self.event_source
            }

            future = self.producer.send(
                self.topic,
                key=user_id,
                value=event
            )
            
            # Wait for the message to be sent (optional, can be async)
            record_metadata = future.get(timeout=10)
            logger.info(
                f'Published license event: {event_type} for user {user_id} '
                f'to topic {record_metadata.topic} partition {record_metadata.partition}'
            )
        except KafkaError as e:
            logger.error(f'Failed to publish license event: {e}')
            raise
        except Exception as e:
            logger.error(f'Unexpected error publishing license event: {e}')
            raise

    def close(self):
        """Close the Kafka producer"""
        if self.producer:
            try:
                self.producer.close()
                logger.info('Kafka producer closed')
            except Exception as e:
                logger.error(f'Error closing Kafka producer: {e}')


# Create singleton instance
kafka_producer = KafkaProducerService()

File: user-service/src/kafka-producer.js
const { Kafka } = require('kafkajs');

class KafkaProducer {
  constructor() {
    const kafkaBroker = process.env.KAFKA_BOOTSTRAP_SERVERS || 'kafka:9093';
    const clientId = process.env.KAFKA_CLIENT_ID || 'user-service';
    this.eventSource = process.env.KAFKA_EVENT_SOURCE || 'user-service-legacy';
    
    this.kafka = new Kafka({
      clientId: clientId,
      brokers: [kafkaBroker],
    });
    
    this.producer = this.kafka.producer();
    this.topic = process.env.KAFKA_USER_EVENTS_TOPIC || 'user-events';
    this.isConnected = false;
  }

  async connect() {
    if (this.isConnected) {
      return;
    }
    
    try {
      await this.producer.connect();
      this.isConnected = true;
      console.log('Kafka producer connected');
    } catch (error) {
      console.error('Failed to connect Kafka producer:', error);
      throw error;
    }
  }

  async disconnect() {
    if (!this.isConnected) {
      return;
    }
    
    try {
      await this.producer.disconnect();
      this.isConnected = false;
      console.log('Kafka producer disconnected');
    } catch (error) {
      console.error('Failed to disconnect Kafka producer:', error);
    }
  }

  async publishEvent(eventType, userId, payload) {
    try {
      if (!this.isConnected) {
        await this.connect();
      }

      const event = {
        eventType,
        userId,
        payload,
        timestamp: new Date().toISOString(),
        source: this.eventSource
      };

      const message = {
        topic: this.topic,
        messages: [{
          key: userId,
          value: JSON.stringify(event)
        }]
      };

      await this.producer.send(message);
      console.log(`Published user event: ${eventType} for user ${userId}`);
    } catch (error) {
      console.error(`Failed to publish user event: ${error.message}`, error);
      throw error;
    }
  }
}

// Create singleton instance
const kafkaProducer = new KafkaProducer();

// Connect on module load
kafkaProducer.connect().catch(err => {
  console.error('Initial Kafka connection failed:', err);
});

// Graceful shutdown
process.on('SIGTERM', async () => {
  await kafkaProducer.disconnect();
});

process.on('SIGINT', async () => {
  await kafkaProducer.disconnect();
  process.exit(0);
});

module.exports = kafkaProducer;

